packages/discord.js/src/client/Client.js

```diff
@@ -231,13 +273,150 @@ class Client extends BaseClient {
+      this.status = Status.Ready;
+
+      this._triggerClientReady();
+      return;
+    }
+    const hasGuildsIntent = this.options.intents.has(GatewayIntentBits.Guilds);
+    // Step 2. Create a timeout that will mark the client as ready if there are still unavailable guilds
+    // * The timeout is 15 seconds by default
+    // * This can be optionally changed in the client options via the `waitGuildTimeout` option
+    // * a timeout time of zero will skip this timeout, which potentially could cause the Client to miss guilds.
+
+    this.readyTimeout = setTimeout(
+      () => {
+        this.emit(
+          Events.Debug,
+          `${
+            hasGuildsIntent
+              ? `Client did not receive any guild packets in ${this.options.waitGuildTimeout} ms.`
+              : 'Client will not receive anymore guild packets.'
+          }\nUnavailable guild count: ${this.expectedGuilds.size}`,
+        );
+
+        this.readyTimeout = null;
+        this.status = Status.Ready;
+
+        this._triggerClientReady();
+      },
+      hasGuildsIntent ? this.options.waitGuildTimeout : 0,
+    ).unref();
+  }
+
+  /**
+   * Attaches event handlers to the WebSocketShardManager from `@discordjs/ws`.
+   * @private
+   */
+  _attachEvents() {
+    this.ws.on(WebSocketShardEvents.Debug, (message, shardId) =>
+      this.emit(Events.Debug, `[WS => ${typeof shardId === 'number' ? `Shard ${shardId}` : 'Manager'}] ${message}`),
+    );
+    this.ws.on(WebSocketShardEvents.Dispatch, this._handlePacket.bind(this));
+
+    this.ws.on(WebSocketShardEvents.Ready, data => {
+      for (const guild of data.guilds) {
+        this.expectedGuilds.add(guild.id);
+      }
+      this.status = Status.WaitingForGuilds;
+      this._checkReady();
+    });
+
+    this.ws.on(WebSocketShardEvents.HeartbeatComplete, ({ heartbeatAt, latency }, shardId) => {
+      this.emit(Events.Debug, `[WS => Shard ${shardId}] Heartbeat acknowledged, latency of ${latency}ms.`);
+      this.lastPingTimestamps.set(shardId, heartbeatAt);
+      this.pings.set(shardId, latency);
+    });
+  }
+
+  /**
+   * Processes a packet and queues it if this WebSocketManager is not ready.
+   * @param {Object} [packet] The packet to be handled
+   * @param {number} [shardId] The shardId that received this packet
+   * @private
+   */
+  _handlePacket(packet, shardId) {
+    if (packet && this.status !== Status.Ready && !BeforeReadyWhitelist.includes(packet.t)) {
+      this.incomingPacketQueue.push({ packet, shardId });
+    } else {
+      if (this.incomingPacketQueue.length) {
+        const item = this.incomingPacketQueue.shift();
+        setImmediate(() => {
+          this._handlePacket(item.packet, item.shardId);
+        }).unref();
+      }
+
+      if (packet && PacketHandlers[packet.t]) {
+        PacketHandlers[packet.t](this, packet, shardId);
+      }
+
+      if (this.status === Status.WaitingForGuilds && WaitingForGuildEvents.includes(packet.t)) {
```

Comment:
Considering `packet` is optional in the method's definition, and is checked accordingly for truthiness in all other branches, this line may throw an error were `packet` to be nullish.